### ISSUE TITLE
PR #4106 causes failures in auto-detecting some machine names

### ISSUE THREAD

**Issue by billsacks** [ESMCI/cime#4133]
The changes in #4106 cause failures in auto-detecting some machine names, resulting in the following (at least for CESM):

```
ERROR: No machine $ENV{NERSC_HOST}:perlmutter found
```

The problem arises when the first call to `_probe_machine_name_one_guess` from `probe_machine_name` fails (because `socket.getfqdn()` doesn't provide the match, but the later call to `socket.gethostname()` will provide the match). The specific issue is that the `machine` variable was used for multiple purposes in the changes to `machines.py` in #4106, so when `_probe_machine_name_one_guess` was supposed to return `None`, it was instead returning the text value of `NODENAME_REGEX` from the last machine with an `$ENV{XXX}:YYY` pattern.

I have a fix on a branch and will open a PR shortly.

**Comment by billsacks** [comment]
I ran into this issue on my Mac. It is *not* an issue on cheyenne (where `socket.getfqdn` provides the necessary name); I haven't tested elsewhere.

**Comment by rljacob** [comment]
@ndkeen noticed that perlmutter only identified itself as "login" which is kind of generic.

**Comment by jedwards4b** [comment]
perlmutter is the reason that this modification was introduced.  We use an environment variable, set on the system NERSC_HOST, to identify the system.

**Comment by ndkeen** [comment]
What should be the NODENAME_REGEX for perlmutter?

```
<NODENAME_REGEX>$ENV{NERSC_HOST}</NODENAME_REGEX>
```

?

**Comment by jedwards4b** [comment]
Not quite, perlmutter doesn't provide a proper hostname through the hostname command we use to identify machines.   We reported the issue to nersc but they didn't seem too eager to fix it.  So we added this feature to cime to collect the hostname from an environment variable.   <NODENAME_REGEX>$ENV{NERSC_HOST}:perlmutter</NODENAME_REGEX>
Using this format the env variable is compared to the string following the ':'
